### PR TITLE
Tutorials

### PR DESCRIPTION
**Pull request by lesleyjanenorton**
## Description

https://user-images.githubusercontent.com/22355127/171444092-e798b646-31fe-4a60-8787-c2b1718c1a56.mov

<table>
<tr>
<td>
Screen Shot 2022-05-31 at 9 59 17 PM
</td>
<td>
Screen Shot 2022-06-01 at 10 54 22 AM

</td>

</tr>
</table>

<ul>
<li>
Finalizes tutorial tooltip UI
</li>
<li>Adds tutorial completed modal</li>

- Clicking 'Let's go!' takes the user back to Tips & Tricks
- Clicking come back later or the X button closes the modal

<li>Adds leave tutorial modal</li>
<li>New tutorial behavior:

- Focus and tabbing is restricted to the tutorial tooltip and the target qml element.
- Opening 'Preferences', 'Contact support', or 'View logs' from the system tray while in the middle of a tutorial will first open a 'Leave tutorial?' modal. Clicking 'Come back later' from that modal will take the user to whichever view they selected from the system tray.
- Tutorials will automatically be stopped if a Captive Portal or Server Unavailable alert needs to be shown.
</li>
</ul>

## Checklist
    
- [ ] My code follows the style guidelines for this project
- [ ] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [ ] I have performed a self review of my own code
- [ ] I have commented my code PARTICULARLY in hard to understand areas
- [ ] I have added thorough tests where needed

### CODE REVIEW

File: nebula/ui/components/VPNControllerView.qml [hunk 1]
```diff
@@ -580,7 +580,7 @@ Item {
                 return;
             }
             Sample.settingsViewOpened.record();
-            VPN.settingsNeeded();
+            mainStackView.push("qrc:/ui/views/ViewSettings.qml", StackView.Immediate)
```
Comment (bakulf) [inline nebula/ui/components/VPNControllerView.qml:583]: why this change?

File: nebula/ui/components/VPNServerList.qml [hunk 2]
```diff
@@ -131,7 +131,7 @@ FocusScope {
                 anchors.right: parent.right
                 anchors.rightMargin: VPNTheme.theme.windowMargin / 2
                 anchors.leftMargin: anchors.rightMargin
-                visible: showRecentConnections && serverSearchInput.text.length === 0
+                visible: showRecentConnections && serverSearchInput.text.length === 0 && !tutorialUI.visible // TODO (unhide recent connections during tutorials)
```
Comment (bakulf) [inline nebula/ui/components/VPNServerList.qml:134]: Can you tell me more about this TODO? Is it a missing thing in the tutorial logic?

File: nebula/ui/components/VPNTutorialPopups.qml [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+import Mozilla.VPN 1.0
+import compat 0.1
+
+
+Item {
+
+    /*
+        TODOs
+
+        - Handle banner alerts
```
Comment (bakulf) [inline nebula/ui/components/VPNTutorialPopups.qml:18]: We can maybe "stop" alerts and update checks during a tutorial

File: nebula/ui/components/VPNTutorialPopups.qml [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+        - Handle conflicts with update required
+
+        - Should tutorials timeout?
+        - Handle device orientation changes
+        - Better UI in landscape mode
+        - Respond to targetElement position changes
+        - Animation polish on tooltip open
+
+        - Update event filter
+            - Add touch events for mobile
```
Comment (bakulf) [inline nebula/ui/components/VPNTutorialPopups.qml:28]: done :)

File: nebula/ui/components/VPNTutorialPopups.qml [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+            - Restrict wheel events
```
Comment (bakulf) [inline nebula/ui/components/VPNTutorialPopups.qml:29]: Do we want to disable them?

File: nebula/ui/components/VPNControllerView.qml (hunk 1 above)
Comment (lesleyjanenorton) [reply nebula/ui/components/VPNControllerView.qml:583]: Because if a tutorial is playing and we call VPN.settingsNeeded(), the 'leave tutorial?' modal will be triggered. Maybe it would be better to add an optional param to `onSettingsNeeded()` to distinguish between calls from clicks to the Settings button and opens from the top bar menu?

File: nebula/ui/components/VPNTutorialPopups.qml (hunk 5 above)
Comment (lesleyjanenorton) [reply nebula/ui/components/VPNTutorialPopups.qml:29]: Ideally yes but as you know, they're stubborn!

File: nebula/ui/components/VPNTutorialPopups.qml (hunk 4 above)
Comment (lesleyjanenorton) [reply nebula/ui/components/VPNTutorialPopups.qml:28]: thankyou! 🥳

File: nebula/ui/components/VPNServerList.qml (hunk 2 above)
Comment (lesleyjanenorton) [reply nebula/ui/components/VPNServerList.qml]: Gone :)

File: src/ui/tutorials/01_get_started.json [hunk 6]
```diff
@@ -41,11 +41,28 @@
...
       "next": {
         "op": "signal",
-        "qml_emitter": "serverCountryList/serverCountry-au/serverCityList/serverCity-Melbourne",
+        "qml_emitter": "serverCountryList/serverCountry-au/serverCityList/serverCity-Melbourne/radioIndicator",
+        "signal": "visibleChanged"
+      }
+    },
+    {
+      "id": "cityAT",
+      "element": "serverCountryList/serverCountry-at/serverCityList/serverCity-Vienna/radioIndicator",
```
[the remark this answers is not in the document]
Comment (lesleyjanenorton) [reply src/ui/tutorials/01_get_started.json]: removed

File: nebula/ui/components/VPNControllerView.qml (hunk 1 above)
Comment (bakulf) [reply nebula/ui/components/VPNControllerView.qml:583]: Yes! This is a good idea. Let's add the parameter.

File: nebula/ui/components/VPNTutorialPopups.qml [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+
+Item {
+
+    /*
+        TODOs
+
+        - Handle banner alerts
+        - Handle conflicts with update required
+
+        - Should tutorials timeout?
```
Comment (bakulf) [inline nebula/ui/components/VPNTutorialPopups.qml:21]: we should ask Peter. It's a very good point!

File: nebula/ui/components/VPNTutorialPopups.qml [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+        - Better UI in landscape mode
+        - Respond to targetElement y position changes
```
Comment (bakulf) [inline nebula/ui/components/VPNTutorialPopups.qml:23]: would it be good if the C++ layer emits a new signal when the 'y' position changes? Or do you want to do it in QML? And... can it be done by QT connecting properties?

File: nebula/ui/components/VPNTutorialPopups.qml [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+
+        id: tutorialTooltip
+
+        closePolicy: Popup.NoAutoClose
+        onClipChanged: VPNTutorial.stop();
+        verticalPadding: VPNTheme.theme.windowMargin
+        horizontalPadding: VPNTheme.theme.windowMargin
+        focus: true
+
+        y: {
```
Comment (bakulf) [inline nebula/ui/components/VPNTutorialPopups.qml:60]: Should we use `Qt.binding(function() { ...` ?

**Review by bakulf — APPROVED** [review]

**Review by MattLichtenstein — CHANGES_REQUESTED** [review]

**Review by MattLichtenstein — APPROVED** [review]

**Review by flodolo — APPROVED** [review]

File: nebula/ui/components/VPNTutorialPopups.qml (hunk 9 above)
Comment (lesleyjanenorton) [reply nebula/ui/components/VPNTutorialPopups.qml:60]: Oooh good question. This may fix the `Respond to targetElement y position changes` TODO.

File: nebula/ui/components/VPNControllerView.qml (hunk 1 above)
Comment (lesleyjanenorton) [reply nebula/ui/components/VPNControllerView.qml:583]: Going to merge this while tests are green and will do this in a fast (later today) follow to avoid touching a bunch of new files.
